units/remote-cryptsetup.target

```diff
@@ -10,7 +10,7 @@
 [Unit]
 Description=Remote Encrypted Volumes
 Documentation=man:systemd.special(7)
-After=remote-fs-pre.target cryptsetup-pre.target
+After=cryptsetup-pre.target
```

Comment:
hmm, so this was borked in 362c378291e85df3e00aaad491d1e08233ad127f (cc @fbuihuu)

i am tempted to just revert 362c378291e85df3e00aaad491d1e08233ad127f. It has the major drawback anyway that it globally orderers all cryptsetup volumes before all mounts. people have stacked stuff though, such a global ordering is just sloppy and problematic in those cases.

I think we should handle the cryptsetup ordering problem that 362c378291e85df3e00aaad491d1e08233ad127f attempted to fix differently. let's stick to local ordering. i.e. introduce an instantiated target unit that is ordered before any mount unit that is backed by a device matching /dev/mapper/*, maybe called mount-prepare@.target. The target would be pulled in by cryptsetup@.service, which would order itself before that.

let's say we have an encrypted volume /dev/mapper/mysql which is ultimately mounted to /var/lib/mysql. For that fstab-generator would automatically synthesize based on some entry /etc/fstab a unit essentially like this:

```
# var-lib-mysql.mount
[Unit]
After=mount-prepare@mysql.target
…

[Mount]
What=/dev/mapper/mysql
Where=/var/lib/mysql
…
```

And cryptsetup-generator would automatically synthesize based on some entry in /etc/crypttab a unit like this:
```
# systemd-cryptsetup@mysql.service
[Unit]
Wants=mount-prepare@%i.target
Before=mount-prepare@%i.target

[Service]
…
```

And we'd ship a static mount-prepare@.target template:

```
# mount-prepare@.target
[Unit]
Description=Passive Mount Preparation Hook
```

That way we'd always get the following ordering:

```
systemd-cryptsetup@mysql.service → mount-prepare@mysql.target
                                                              🡖 
                                                                  var-lib-mysql.mount
                                                              🡕
                                     dev-mapper-mysql.device
```

With that we make sure the mount is for the first time properly ordered after cryptsetup *and* the device popping up, which is already pretty nice, but more importantly means at shutdown we'd first unmount the thing and only then destroy the luks volume.

Does that make sense?

Reply:
The units already have a "connection" through the .device unit. If these were just normal units, we could simply order them `systemd-cryptsetup@mysql.service Before dev-mapper-mysql.device Before var-lib-mysql.mount`, and we would get correct job ordering on shutdown. I know .device units cannot be delayed, but maybe we could still add an ordering? I'm confused why some .device units have After=/Before= deps, and others don't and when it is allowed to have them.

Reply:
with your ordering at shutdown we'd first umount the mount point then wait for the device to disappear and then stop the cryptsetup service. But that's not desired, we want to stop the service first and that results in the device to disappear, i.e. the device coming/going is a side effect of cryptsetup service and not an event that can be ordered before or after it. 

if you so will it's effect of systemd's design model that shutdown order is the reverse of the startup order. because here during both startup and shutdown we want the invoke cryptsetup first, and that then has effect on the device unit.

Hence I think the right approach to fit this into our model is to consider the .device and .service units as independently ordered but then insert an additional unit in between to order both together against the stuff we care about.

Reply:
> i am tempted to just revert [362c378](https://github.com/systemd/systemd/commit/362c378291e85df3e00aaad491d1e08233ad127f). It has the major drawback anyway that it globally orderers all cryptsetup volumes before all mounts. people have stacked stuff though, such a global ordering is just sloppy and problematic in those cases.

Could you please be more descriptive here ?

Reply:
I am prepping a patch implementing what I propose right now.

Reply:
I posted PR #14398 now (wip) which does this how I think this should be done.